### ISSUE TITLE
NSFileProvider Extension crash on MacOS 14.4 Beta

### ISSUE THREAD

**Issue by msavkin** [xamarin/xamarin-macios#20034]
### Steps to Reproduce

1. Create simple Xamarin.Mac solution and add FileProviderExtension project
2. Storage appears in Locations section in Finder

### Expected Behavior
Cloud storage shows folders content

### Actual Behavior
Cloud storage doesn't show folder content, just infinite running spinner
<img width="735" alt="Screenshot 2024-02-05 at 15 11 18" src="https://github.com/xamarin/xamarin-macios/assets/78610481/45478fb5-670a-4458-ab55-4bc89ae73900">
FileProviderExtension process constantly crashes. It occurs only in OSX starting from 14.4 beta

### Environment

<details>
<summary>Version information</summary>

```
Visual Studio Professional 2022 for Mac
Version 17.6.8 (build 400)
Installation UUID: 8e025732-c7b3-42a0-9f1d-094c086b5732

Runtime
.NET 7.0.3 (64-bit)
Architecture: Arm64
Microsoft.macOS.Sdk 13.1.1007; git-rev-head:8afca776a0a96613dfb7200e0917bb57f9ed5583; git-branch:release/7.0.1xx-xcode14.2

Roslyn (Language Service)
4.6.0-3.23180.6+99e956e42697a6dd886d1e12478ea2b27cceacfa

NuGet
Version: 6.4.0.117

.NET SDK (Arm64)
SDK: /usr/local/share/dotnet/sdk/7.0.312/Sdks
SDK Versions:
	8.0.100
	7.0.312
	7.0.311
	7.0.308
	7.0.306
	7.0.302
	7.0.202
	7.0.200
	6.0.418
	6.0.417
	6.0.414
	6.0.412
	6.0.408
	6.0.407
	6.0.406
	6.0.401
	6.0.301
	6.0.300
MSBuild SDKs: /Applications/Visual Studio.app/Contents/MonoBundle/MSBuild/Current/bin/Sdks

.NET SDK (x64)
SDK Versions:
	6.0.418
	6.0.417
	6.0.414
	6.0.412
	6.0.408
	6.0.407
	6.0.401
	6.0.301
	6.0.105
	6.0.104
	6.0.103
	6.0.102
	6.0.101
	5.0.408
	5.0.407
	5.0.406
	5.0.405
	5.0.404
	3.1.426
	3.1.423
	3.1.420
	3.1.419
	3.1.418
	3.1.417
	3.1.416

.NET Runtime (Arm64)
Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	8.0.0
	7.0.15
	7.0.14
	7.0.11
	7.0.9
	7.0.5
	7.0.4
	7.0.3
	6.0.26
	6.0.25
	6.0.22
	6.0.20
	6.0.16
	6.0.15
	6.0.14
	6.0.9
	6.0.6
	6.0.5

.NET Runtime (x64)
Runtime: /usr/local/share/dotnet/x64/dotnet
Runtime Versions:
	6.0.26
	6.0.25
	6.0.22
	6.0.20
	6.0.16
	6.0.15
	6.0.9
	6.0.6
	6.0.5
	6.0.4
	6.0.3
	6.0.2
	6.0.1
	5.0.17
	5.0.16
	5.0.15
	5.0.14
	5.0.13
	3.1.32
	3.1.29
	3.1.26
	3.1.25
	3.1.24
	3.1.23
	3.1.22

Xamarin.Profiler
Version: 1.8.0.49
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

Updater
Version: 11

Apple Developer Tools
Xcode: 15.3 22617
Build: 15E5178i

Xamarin.Mac
Version: 9.3.0.23 Visual Studio Professional
Hash: 9defd91b3
Branch: xcode14.3
Build date: 2023-10-23 16:14:59-0400

Xamarin.iOS
Version: 16.4.0.23 Visual Studio Professional
Hash: 9defd91b3
Branch: xcode14.3
Build date: 2023-10-23 16:15:00-0400

Xamarin Designer
Version: 17.6.3.9
Hash: 2648399ae8
Branch: remotes/origin/d17-6
Build date: 2024-01-03 23:07:21 UTC

Xamarin.Android
Version: 13.2.2.0 (Visual Studio Professional)
Commit: xamarin-android/d17-5/45b0e14
Android SDK: /Users/msavkin/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		13.0 (API level 33)

SDK Command-line Tools Version: 7.0
SDK Platform Tools Version: 34.0.3
SDK Build Tools Version: 32.0.0

Build Information: 
Mono: d9a6e87
Java.Interop: xamarin/java.interop/d17-5@149d70fe
SQLite: xamarin/sqlite/3.40.1@68c69d8
Xamarin.Android Tools: xamarin/xamarin-android-tools/d17-5@ca1552d

Microsoft Build of OpenJDK
Java SDK: /Library/Java/JavaVirtualMachines/microsoft-11.jdk
11.0.16.1
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

Eclipse Temurin JDK
Java SDK: /Library/Java/JavaVirtualMachines/temurin-8.jdk
1.8.0.302
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

Android SDK Manager
Version: 17.6.0.50
Hash: a715dca
Branch: HEAD
Build date: 2024-01-03 23:07:29 UTC

Android Device Manager
Version: 0.0.0.1309
Hash: 06e3e77
Branch: HEAD
Build date: 2024-01-03 23:07:29 UTC

Build Information
Release ID: 1706080400
Git revision: dfc3b79e7cc0dbc2710b239472a704042f66dad0
Build date: 2024-01-03 23:05:33+00
Build branch: release-17.6
Build lane: release-17.6

Operating System
Mac OS X 14.4.0
Darwin 23.4.0 Darwin Kernel Version 23.4.0
    Fri Jan 12 22:41:06 PST 2024
    root:xnu-10063.100.610.0.2~19/RELEASE_ARM64_T8103 arm64
```

</details>

### Example Project (If Possible)

[FP.zip](https://github.com/xamarin/xamarin-macios/files/14165537/FP.zip)

### Example of crash reports
[Workplace.Mac.FileProvider.Extension-2024-02-05-151112.000.ips.zip](https://github.com/xamarin/xamarin-macios/files/14165643/Workplace.Mac.FileProvider.Extension-2024-02-05-151112.000.ips.zip)

Please review crash report and help me understand whats wrong and how to fix it. Simple project written with Swift and Xcode works without any issues. Thanks a lot!

**Comment by emergencynap** [comment]
Same issue with 14.4 beta 2

**Comment by rolfbjarne** [comment]
I'll try to update one of my machines in the upcoming days and see if I can reproduce this.

**Comment by rolfbjarne** [comment]
I can reproduce.

It's a change in macOS: if there's a native `main` function in the extension, they'll now call it. Unfortunately we don't expect that, and things break.

You can easily reproduce in an Xcode project by adding this to the file provider extension:

```c
int main (int argc, char ** argv)
{
    abort ();
    return 2;
}
```

that will abort in macOS 14.4 beta 3, and work in macOS 14.3.

**Comment by rolfbjarne** [comment]
I've filed a feedback ticket with Apple (FB13637117). Feel free to file a ticket yourself as well (I believe Apple prioritizes issues based on number of duplicates, so the more people file the same issue, the likelier they are to look at it / fix it).

**Comment by rolfbjarne** [comment]
(Apple feedback replicated below)

# Apple feedback

## File Provider extensions call any native "main" methods in the extension in macOS 14.4

FB13637117

### Description

If a file provider extension has a native "main" method, that method is now called in macOS 14.4 seed 1-3, while in macOS 14.3 the "main" method is not called.

This is easy to reproduce: add a "main" method that aborts to a file provider extension:

```c
int main (int argc, char ** argv)
{
    abort ();
    return 1;
}
```

And see how it aborts in macOS 14.4 and works in macOS 14.3.

Example crash report attached.

I've also attached a zip file with the test project (FPTestX).

This is a breaking change for file provider extensions that have already been released, they won't work in macOS 14.4.

Test project: [FPTestX-e38de1f.zip](https://github.com/xamarin/xamarin-macios/files/14347650/FPTestX-e38de1f.zip)

**Comment by rolfbjarne** [comment]
I believe this will be somewhat complicated to fix on our side, so we might just wait and see if Apple fixes it on their side.

**Comment by AndirNotes** [comment]
@rolfbjarne Any news? 
Looks like released MacOS Sonoma 14.4 and 14.4.1 crashes as well. 

Additional info I see those logs in Console just before the crash:
```
debug	20:03:42.351394+0400	SampleFileProviderExt	com.apple.extensionkit	default	SwiftUI: Swift entry point data not found.
debug	20:03:42.351507+0400	SampleFileProviderExt	com.apple.extensionkit	default	SwiftUI: Swift entry point address could not be determined.
debug	20:03:42.351561+0400	SampleFileProviderExt	com.apple.extensionkit	default	SwiftUI: Falling back to main function lookup via dlsym().
```
Looks like this main method call is pretty intentional.

Maybe you can suggest any workaround? can we somehow to remove this Xamarin.Mac main from exported symbols?

**Comment by rolfbjarne** [comment]
According to Apple, this should be fixed in the first beta for macOS 15.0

**Comment by rolfbjarne** [comment]
Closing this, because:

1. There's a workaround (https://github.com/xamarin/xamarin-macios/issues/20034#issuecomment-2030155285)
2. Apple says they've fixed it on their end.
3. This only happens with legacy Xamarin, which is out of support now.
